File: C#/Graph - BFS/994. Rotting Oranges.cs
/* 994. Rotting Oranges
Difficulty: Medium
Submission Link: https://leetcode.com/problems/rotting-oranges/submissions/1385732070/ */

public class Solution {
    public const int fresh = 1, rotten = 2;

    public readonly (int, int)[] directions = {
        (0, 1),
        (0, -1),
        (1, 0),
        (-1, 0),
    };

    public int OrangesRotting(int[][] grid) {
        var m = grid.Length;
        var n = grid[0].Length;
        var queue = new Queue<(int, int)>();
        var minutes = 0;
        var numFresh = 0;

        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) 
                switch (grid[i][j])
                {
                    case fresh: 
                        numFresh++;
                        break;
                    case rotten: 
                        queue.Enqueue((i, j));
                        break;
                }
        
        while (queue.Count > 0)
        {
            var boardChanged = false;
            var nextQueue = new Queue<(int, int)>();

            while (queue.Count > 0)
            {
                var (row, col) = queue.Dequeue();

                foreach (var (x, y) in directions)
                {
                    var (newRow, newCol) = (row + x, col + y);

                    if (newRow < 0 || newRow >= m) continue;
                    if (newCol < 0 || newCol >= n) continue;
                    if (grid[newRow][newCol] != fresh) continue;

                    boardChanged = true;
                    grid[newRow][newCol] = rotten;
                    numFresh--;
                    nextQueue.Enqueue((newRow, newCol));
                }
            }

            if (boardChanged) minutes++;
            queue = nextQueue;
        }

        if (numFresh != 0) return -1;

        return minutes;
    }
}

File: Python3/Graph - BFS/994. Rotting Oranges.py
""" 994. Rotting Oranges
Difficulty: Medium
Submission Link: https://leetcode.com/problems/rotting-oranges/submissions/1385731476/ """

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        queue = deque()
        fresh_oranges = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    queue.append((i, j))
                elif grid[i][j] == 1:
                    fresh_oranges += 1
        if fresh_oranges == 0:
            return 0
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        time = 0
        
        while queue:
            time += 1
            for _ in range(len(queue)):
                x, y = queue.popleft()
                for dx, dy in directions:
                    nx, ny = x + dx, y + dy
                    if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 1:
                        grid[nx][ny] = 2
                        queue.append((nx, ny))
                        fresh_oranges -= 1
            if fresh_oranges == 0:
                return time
        return -1